File: parser/src/main/java/org/cooklang/parser/Step.java
package org.cooklang.parser;


import java.util.ArrayList;
import java.util.List;

public class Step {
    public List<DirectionItem> directions = new ArrayList<>();

    public void addTextItem(TextItem texItem) {
        directions.add(texItem);
    }

    public void addIngredient(Ingredient ingredient) {
        directions.add(ingredient);
    }

    public void addCookware(Cookware cookware) {
        directions.add(cookware);
    }

    public void addTimer(Timer timer) {
        directions.add(timer);
    }

    public String toString() {
        if (directions == null || directions.size() <= 0) return "";

        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < directions.size(); i++) {
            sb.append(directions.get(i).toString());
        }

        return sb.toString();
    }
}

File: parser/src/main/java/org/cooklang/parser/Step.kt
package org.cooklang.parser

import org.cooklang.parser.DirectionItem
import org.cooklang.parser.TextItem
import org.cooklang.parser.Ingredient
import org.cooklang.parser.Cookware
import java.lang.StringBuilder
import java.util.ArrayList

class Step: Displayable {
    val directions: MutableList<DirectionItem> = mutableListOf()
    fun addTextItem(texItem: TextItem) {
        directions.add(texItem)
    }

    fun addIngredient(ingredient: Ingredient) {
        directions.add(ingredient)
    }

    fun addCookware(cookware: Cookware) {
        directions.add(cookware)
    }

    fun addTimer(timer: Timer) {
        directions.add(timer)
    }
    val ingredients: List<Ingredient> get() = directions.filterIsInstance(Ingredient::class.java)
    val cookware: List<Cookware> get() = directions.filterIsInstance(Cookware::class.java)

    override fun getDisplayString(): String {
        if (directions.size <= 0) return ""
        val sb = StringBuilder()
        directions.forEach {
            sb.append(it.getDisplayString())
        }
        return sb.toString()
    }
}